File: qualitycheck1.R
# Note Elias Jedam: Python Code translated into R by basically Chatgpt
library(readxl)
library(dplyr)
library(ggplot2)
getwd()
#get data
df <- read_excel("data/Dataset_Assignment_1.xlsx", sheet = 1)

# categorical variables
df <- df %>%
  mutate(
    Gender = recode(Gender,
                    "Male" = 0,
                    "Female" = 1),
    Accommodation_Class = recode(Accommodation_Class,
                                 "Rented" = 0,
                                 "Non_Rented" = 1)
  )

# general check of data quality
colSums(is.na(df))

features <- c("DTI", "FOIR", "LTV", "Age")

#histogram oldemi
hist_info <- hist(df$OldEmi,
                  main = "",
                  xlab = "OldEmi")

text(hist_info$mids,
     hist_info$counts,
     labels = hist_info$counts,
     pos = 3)

# pdf("pictures/table1_OldEmi.pdf")
# hist(...)
# dev.off()

# gender distributions
counts <- table(df$Gender)

male <- counts["0"]
female <- counts["1"]

cat("Male : Female =", male, ":", female, "\n")
cat("Ratio =", round(as.numeric(male/female), 2), "\n")

# AC distribution
counts <- table(df$Accommodation_Class)

rented <- counts["0"]
non_rented <- counts["1"]

cat("Rented : Non Rented =", rented, ":", non_rented, "\n")
cat("Ratio =", round(as.numeric(rented/non_rented), 2), "\n")

#histogram of all features
features <- c(
  "DTI", "FOIR", "LTV",
  "Age", "OldEmi",
  "Gender", "Accommodation_Class"
)

par(mfrow = c(3,3))

for(feature in features){
  hist(df[[feature]],
       main = feature,
       xlab = feature,
       col = "lightgray")
}

par(mfrow = c(1,1))

#boxplots
par(mfrow = c(2,4))

for(feature in features){
  boxplot(df[[feature]],
          main = feature)
}

par(mfrow = c(1,1))


prop.table(table(df$Gender, df$Decision), margin = 1)
prop.table(table(df$Accommodation_Class, df$Decision), margin = 1)

# Countplots
ggplot(df, aes(x = factor(Gender), fill = Decision)) +
  geom_bar(position = "dodge") +
  labs(x = "Gender")

ggplot(df, aes(x = factor(Accommodation_Class), fill = Decision)) +
  geom_bar(position = "dodge") +
  labs(x = "Accommodation Class")

#KDE plots
features <- c("DTI","FOIR","LTV","Age","OldEmi")

par(mfrow = c(2,3))

for(feature in features){
  
  plot(
    density(df[df$Decision=="Approve", feature][[1]]),
    main = feature,
    xlab = feature,
    col = "blue",
    lwd = 2
  )
  
  lines(
    density(df[df$Decision=="Reject", feature][[1]]),
    col = "red",
    lwd = 2
  )
  
  legend("topright",
         legend = c("Approve","Reject"),
         col = c("blue","red"),
         lwd = 2)
}

# Zoom OldEmi
plot(
  density(df[df$Decision=="Approve","OldEmi"][[1]]),
  xlim = c(20000,40000),
  main = "OldEmi (Zoom)",
  xlab = "OldEmi",
  col = "blue",
  lwd = 2
)

lines(
  density(df[df$Decision=="Reject","OldEmi"][[1]]),
  col = "red",
  lwd = 2
)

legend("topright",
       legend = c("Approve","Reject"),
       col = c("blue","red"),
       lwd = 2)

par(mfrow = c(1,1))

#histogram seperated by decision
par(mfrow = c(2,3))

for(feature in features){
  
  hist(df[df$Decision=="Approve", feature][[1]],
       breaks = 20,
       col = rgb(0,0,1,0.4),
       main = feature,
       xlab = feature)
  
  hist(df[df$Decision=="Reject", feature][[1]],
       breaks = 20,
       col = rgb(1,0,0,0.4),
       add = TRUE)
  
  legend("topright",
         legend = c("Approve","Reject"),
         fill = c(rgb(0,0,1,0.4), rgb(1,0,0,0.4)))
}

par(mfrow = c(1,1))

#histogram oldemi
hist(df[df$Decision=="Approve","OldEmi"][[1]],
     breaks = 20,
     col = rgb(0,0,1,0.4),
     xlab = "OldEmi",
     main = "OldEmi")

hist(df[df$Decision=="Reject","OldEmi"][[1]],
     breaks = 20,
     col = rgb(1,0,0,0.4),
     add = TRUE)

legend("topright",
       legend = c("Approve","Reject"),
       fill = c(rgb(0,0,1,0.4), rgb(1,0,0,0.4)))

File: qualitycheck1.py
# %%
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

df = pd.read_excel("data/Dataset_Assignment_1.xlsx", sheet_name=0)
df["Gender"] = df["Gender"].map({"Male": 0, "Female": 1})
df["Accommodation_Class"] = df["Accommodation_Class"].map(
    {"Rented": 0, "Non_Rented": 1}
)

# %%
print(df.isnull().sum())
features = ["DTI", "FOIR", "LTV", "Age"]

counts, bins, patches = plt.hist(df["OldEmi"])
for count, patch in zip(counts, patches):
    plt.text(
        patch.get_x() + patch.get_width() / 2,
        count,
        int(count),
        ha="center",
        va="bottom",
    )
plt.savefig("pictures/table1_OldEmi.pdf", bbox_inches="tight")
plt.show()

counts = df["Gender"].value_counts()
male = counts[0]  # male
female = counts[1]
print(f"Male : Female = {male}:{female}")
print(f"Ratio = {male/female:.2f}")

counts = df["Accommodation_Class"].value_counts()
male = counts[0]  # rented
female = counts[1]
print(f"Rented : Non Rented = {male}:{female}")
print(f"Ratio = {male/female:.2f}")


# %%
features = ["DTI", "FOIR", "LTV", "Age", "OldEmi", "Gender", "Accommodation_Class"]

df[features].hist(figsize=(12, 8), bins=30)
plt.tight_layout()
plt.savefig("pictures/histograms.pdf", bbox_inches="tight")


# %%
plt.figure(figsize=(10, 5))

for i, feature in enumerate(features):
    plt.subplot(2, 4, i + 1)
    plt.boxplot(df[feature])
    plt.title(feature)

plt.tight_layout()
plt.savefig("pictures/boxplots.pdf", bbox_inches="tight")

# %%
print(pd.crosstab(df["Gender"], df["Decision"], normalize="index"))

# %%
print(pd.crosstab(df["Accommodation_Class"], df["Decision"], normalize="index"))

# %%

sns.countplot(data=df, x="Gender", hue="Decision")
plt.show()

sns.countplot(data=df, x="Accommodation_Class", hue="Decision")
plt.show()

# %%

features = ["DTI", "FOIR", "LTV", "Age", "OldEmi"]

fig, axes = plt.subplots(2, 3, figsize=(15, 8))
axes = axes.flatten()

# First five KDE plots
for i, feature in enumerate(features):
    sns.kdeplot(
        data=df,
        x=feature,
        hue="Decision",
        common_norm=False,
        fill=True,
        alpha=0.3,
        ax=axes[i],
    )

    axes[i].set_title(feature)
    axes[i].set_xlabel(feature)
    axes[i].set_ylabel("Density")

# Sixth plot: zoomed OldEmi
sns.kdeplot(
    data=df,
    x="OldEmi",
    hue="Decision",
    common_norm=False,
    fill=True,
    alpha=0.3,
    ax=axes[5],
)

axes[5].set_xlim(20000, 40000)
axes[5].set_ylim(0, 0.000006)
axes[5].set_title("OldEmi (Zoom)")
axes[5].set_xlabel("OldEmi")
axes[5].set_ylabel("Density")

plt.tight_layout()
plt.savefig("pictures/densities.pdf", bbox_inches="tight")
plt.show()

# %%
features = ["DTI", "FOIR", "LTV", "Age", "OldEmi"]

fig, axes = plt.subplots(2, 3, figsize=(12, 8))

for ax, feature in zip(axes.flat, features):

    ax.hist(
        df[df["Decision"] == "Approve"][feature], bins=20, alpha=0.6, label="Approve"
    )

    ax.hist(df[df["Decision"] == "Reject"][feature], bins=20, alpha=0.6, label="Reject")

    ax.set_title(feature)

axes[0, 0].legend()
plt.tight_layout()
plt.show()

# %%
import matplotlib.pyplot as plt

plt.figure(figsize=(6, 4))

plt.hist(df[df["Decision"] == "Approve"]["OldEmi"], bins=20, alpha=0.6, label="Approve")

plt.hist(df[df["Decision"] == "Reject"]["OldEmi"], bins=20, alpha=0.6, label="Reject")

plt.xlabel("FOIR")
plt.ylabel("Count")
plt.legend()
plt.show()

# %%
